### PR TITLE
[shopsys] Unify working with images

### PR DESCRIPTION
**Pull request by malyMiso**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Working with single image is the same as working with multiple images. There exists a new function `ImageFacade::manageImages`, which unites the calling of the orginal methods. It gives an advantage to easily switch between multiple and single image support for the entities. Moreover, it allows admin to delete single image.
|New feature| No
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| Yes
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions]| Yes

### CODE REVIEW

File: packages/framework/src/Component/Image/ImageFacade.php [hunk 1]
```diff
@@ -92,10 +93,36 @@ public function __construct(
...
+        }
+
+        $this->deleteImages($entity, $imageUploadData->imagesToDelete);
+    }
+
     /**
      * @param object $entity
      * @param array $temporaryFilenames
      * @param string|null $type
+     * @deprecated This function is deprecated since 7.3.2, use ImageFacade::manageImages instead
```
Comment (DavidOstrozlik) [inline packages/framework/src/Component/Image/ImageFacade.php]: about the message - Is it really deprecated or will it be changed to protected in the next release?

File: packages/framework/src/Form/Admin/Advert/AdvertFormType.php [hunk 2]
```diff
@@ -39,16 +40,24 @@ class AdvertFormType extends AbstractType
...
     /**
      * @param \Shopsys\FrameworkBundle\Component\Domain\Domain $domain
      * @param \Shopsys\FrameworkBundle\Model\Advert\AdvertPositionRegistry $advertPositionRegistry
+     * @param \Shopsys\FrameworkBundle\Component\Image\Config\ImageConfig $imageConfig
      */
     public function __construct(
         Domain $domain,
-        AdvertPositionRegistry $advertPositionRegistry
+        AdvertPositionRegistry $advertPositionRegistry,
+        ImageConfig $imageConfig
```
Comment (DavidOstrozlik) [inline packages/framework/src/Form/Admin/Advert/AdvertFormType.php]: Adding a new parameter to a constructor is a BC break according to our BC promise.
You can make this nullable with default null value and add setter injection.
```php
    ?ImageConfig $imageConfig = null 
```

``` php
/**
     * @required
     * @internal This function will be replaced by constructor injection in next major
     * @param \Shopsys\FrameworkBundle\Component\Cron\CronModuleExecutor $cronModuleExecutor
     */
    public function setCronModuleExecutor(CronModuleExecutor $cronModuleExecutor)
    {
        if ($this->cronModuleExecutor !== null && $this->cronModuleExecutor !== $cronModuleExecutor) {
            throw new BadMethodCallException(sprintf('Method "%s" has been already called and cannot be called multiple times.', __METHOD__));
        }
        if ($this->cronModuleExecutor === null) {
            @trigger_error(sprintf('The %s() method is deprecated and will be removed in the next major. Use the constructor injection instead.', __METHOD__), E_USER_DEPRECATED);
            $this->cronModuleExecutor = $cronModuleExecutor;
        }
    }
```

File: packages/framework/src/Component/Image/ImageFacade.php (hunk 1 above)
Comment (malyMiso) [reply packages/framework/src/Component/Image/ImageFacade.php]: Hi, thx. Fixed

File: packages/framework/src/Form/Admin/Advert/AdvertFormType.php (hunk 2 above)
Comment (malyMiso) [reply packages/framework/src/Form/Admin/Advert/AdvertFormType.php]: Thx. DI has been change to setter

File: packages/framework/src/Form/Admin/Product/Brand/BrandFormType.php [hunk 3]
```diff
@@ -32,16 +34,40 @@ class BrandFormType extends AbstractType
...
+        SeoSettingFacade $seoSettingFacade,
+        ?ImageConfig $imageConfig = null
     ) {
         $this->domain = $domain;
         $this->seoSettingFacade = $seoSettingFacade;
+        $this->imageConfig = $imageConfig;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Form/Admin/Product/Brand/BrandFormType.php]:
```suggestion
     * @required
```

File: packages/framework/src/Form/Admin/Product/ProductFormType.php [hunk 4]
```diff
@@ -157,6 +166,23 @@ public function __construct(
         $this->pricingGroupFacade = $pricingGroupFacade;
         $this->pluginDataFormExtensionFacade = $pluginDataFormExtensionFacade;
         $this->productParameterValueToProductParameterValuesLocalizedTransformer = $productParameterValueToProductParameterValuesLocalizedTransformer;
+        $this->imageConfig = $imageConfig;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Form/Admin/Product/ProductFormType.php]:
```suggestion
     * @required
```

File: packages/framework/src/Form/Admin/Slider/SliderItemFormType.php [hunk 5]
```diff
@@ -22,6 +24,35 @@ class SliderItemFormType extends AbstractType
...
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Image\Config\ImageConfig|null $imageConfig
+     */
+    public function __construct(?ImageConfig $imageConfig = null)
+    {
+        $this->imageConfig = $imageConfig;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Form/Admin/Slider/SliderItemFormType.php]:
```suggestion
     * @required
```

File: packages/framework/src/Form/Admin/Transport/TransportFormType.php [hunk 6]
```diff
@@ -46,22 +48,46 @@ class TransportFormType extends AbstractType
...
     ) {
         $this->vatFacade = $vatFacade;
         $this->paymentFacade = $paymentFacade;
         $this->currencyFacade = $currencyFacade;
         $this->transportFacade = $transportFacade;
+        $this->imageConfig = $imageConfig;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Form/Admin/Transport/TransportFormType.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Advert/AdvertDataFactory.php [hunk 7]
```diff
@@ -2,8 +2,40 @@
...
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Image\ImageFacade|null $imageFacade
+     */
+    public function __construct(?ImageFacade $imageFacade = null)
+    {
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Advert/AdvertDataFactory.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Category/CategoryDataFactory.php [hunk 8]
```diff
@@ -28,22 +30,46 @@ class CategoryDataFactory implements CategoryDataFactoryInterface
...
     ) {
         $this->categoryRepository = $categoryRepository;
         $this->friendlyUrlFacade = $friendlyUrlFacade;
         $this->pluginCrudExtensionFacade = $pluginCrudExtensionFacade;
         $this->domain = $domain;
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Category/CategoryDataFactory.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Payment/PaymentDataFactory.php [hunk 9]
```diff
@@ -22,19 +24,43 @@ class PaymentDataFactory implements PaymentDataFactoryInterface
...
+        ?ImageFacade $imageFacade = null
     ) {
         $this->paymentFacade = $paymentFacade;
         $this->vatFacade = $vatFacade;
         $this->domain = $domain;
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Payment/PaymentDataFactory.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Product/Brand/BrandDataFactory.php [hunk 10]
```diff
@@ -22,19 +24,43 @@ class BrandDataFactory implements BrandDataFactoryInterface
...
+        ?ImageFacade $imageFacade = null
     ) {
         $this->friendlyUrlFacade = $friendlyUrlFacade;
         $this->brandFacade = $brandFacade;
         $this->domain = $domain;
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Product/Brand/BrandDataFactory.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Slider/SliderItemDataFactory.php [hunk 11]
```diff
@@ -2,8 +2,40 @@
...
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Image\ImageFacade|null $imageFacade
+     */
+    public function __construct(?ImageFacade $imageFacade = null)
+    {
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Slider/SliderItemDataFactory.php]:
```suggestion
     * @required
```

File: packages/framework/src/Model/Transport/TransportDataFactory.php [hunk 12]
```diff
@@ -22,19 +24,43 @@ class TransportDataFactory implements TransportDataFactoryInterface
...
+        ?ImageFacade $imageFacade = null
     ) {
         $this->transportFacade = $transportFacade;
         $this->vatFacade = $vatFacade;
         $this->domain = $domain;
+        $this->imageFacade = $imageFacade;
+    }
+
+    /**
+     * @reqire
```
Comment (DavidOstrozlik) [inline packages/framework/src/Model/Transport/TransportDataFactory.php]:
```suggestion
     * @required
```

File: upgrade/UPGRADE-v8.1.0-dev.md [hunk 13]
```diff
@@ -371,6 +371,18 @@ There you can find links to upgrade notes for other versions too.
         +        Assert::assertNotSame($errorIdFirstAccess, $errorIdSecondAccess);
         +    }
         ```
+- autowire service `ImageFacade` in data object factories ([#1476](https://github.com/shopsys/shopsys/pull/1476))
+    - `src/Shopsys/ShopBundle/Model/Payment/PaymentDataFactory.php`
+    ```diff
```
Comment (DavidOstrozlik) [inline upgrade/UPGRADE-v8.1.0-dev.md]: Can you please add to this diff the whole change of the constructor?
```diff
    public function __construct(
        PaymentFacade $paymentFacade,
        VatFacade $vatFacade,
-       Domain $domain
+       Domain $domain,
+       ImageFacade $imageFacade
    ) {
-       parent::__construct($paymentFacade, $vatFacade, $domain);
+       parent::__construct($paymentFacade, $vatFacade, $domain, $imageFacade);
```

Comment (malyMiso) [reply upgrade/UPGRADE-v8.1.0-dev.md]: Thx. fixed

File: upgrade/UPGRADE-v8.1.0-dev.md [hunk 14]
```diff
@@ -371,6 +371,33 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+            public function __construct(
+                PaymentFacade $paymentFacade,
+                VatFacade $vatFacade,
+        -       Domain $domain
+        +       Domain $domain,
+        +       ImageFacade $imageFacade
+            ) {
+        -       parent::__construct($paymentFacade, $vatFacade, $domain);
+        +       parent::__construct($paymentFacade, $va
```
Comment (DavidOstrozlik) [inline upgrade/UPGRADE-v8.1.0-dev.md]: You are missing something here :smile: 
```suggestion
        +       parent::__construct($paymentFacade, $vatFacade, $domain, $imageFacade);
```

**Pull request merged**